packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -998,7 +999,38 @@ function useMutableSource<Source, Snapshot>(
 
   // Sync the values needed by our subscribe function after each commit.
   dispatcher.useEffect(() => {
+    const didGetSnapshotChange = !is(refs.getSnapshot, getSnapshot);
     refs.getSnapshot = getSnapshot;
+
+    // Normally the dispatch function for a state hook never changes,
+    // but in the case of this hook, it will change if getSnapshot changes.
+    // In that case, the subscription below will have cloesd over the previous function,
+    // so we use a ref to ensure that handleChange() always has the latest version.
+    refs.setSnapshot = setSnapshot;
+
+    // This effect runs on mount, even though getSnapshot hasn't changed.
+    // In that case we can avoid the additional checks for a changed snapshot,
+    // because the subscription effect below will cover this.
+    if (didGetSnapshotChange) {
+      // Because getSnapshot is shared with subscriptions via a ref,
+      // we don't resubscribe when getSnapshot changes.
+      // This means that we also don't check for any missed mutations
+      // between the render and the passive commit though.
+      // So we need to check here, just like when we newly subscribe.
+      const maybeNewVersion = getVersion(source._source);
+      if (!is(version, maybeNewVersion)) {
+        const maybeNewSnapshot = getSnapshot(source._source);
+        if (!is(snapshot, maybeNewSnapshot)) {
+          // If the source mutated between render and now,
+          // we may have a state update already scheduled from the old getSnapshot.
+          // It's important to ensure that this update is included with that one.
+          // TODO This sucks; what's the right way of doing this?
+          runWithPriority(UserBlockingPriority, () => {
+            setSnapshot(maybeNewSnapshot);
```

Comment:
As I mentioned [in this comment](https://github.com/facebook/react/pull/18296#issuecomment-600238790), I think what I need to do here is schedule a state update with the same expiration time as the one that might be pending due to a mutation. (Either that or manually replace the pending update with this one, but that seemed a little shady too.)

Added this "fix" in the meanwhile, for discussion purposes. :smile: